### DISCUSSION TITLE
Retry on GitHub secondary rate limits

### DISCUSSION THREAD

**Discussion by oschwald** [renovatebot/renovate#24825] [Suggest an Idea]
### What would you like Renovate to be able to do?

GitHub appears to use a 403 status code with the following body for (some) rate limiting:

```json
{"documentation_url":"https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits","message":"You have exceeded a secondary rate limit. Please wait a few minutes before you try again."}
```

There seems to be some special code for handling this response already:

https://github.com/renovatebot/renovate/blob/e132c34c53c351be22e9459988a6d8cc190ddb16/lib/util/http/github.ts#L93-L99

However, it would preferable if this was treated as a 429 and retried. The response does have a `Retry-After` header on it, which is set to 60 in the instances I have seen.

Currently, when the 403 is hit, that request and requests immediately following it will fail. This appears to result in Renovate closing any open PRs for the dependency.

### If you have any ideas on how this should be implemented, please tell us here.

The request should be retries after `Retry-After`.

### Is this a feature you are interested in implementing yourself?

No

**Comment by viceice** [comment]
i think we somehow should skip that dep in this run, as the retry time can be long and we don't want to block unnecessary.

value can be a date or seconds.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After

**Comment by rarkins** [comment]
Skipping the dep instead of the whole run can lead to autoclosings/flip-floppings. Maybe we allow a maximum retry value to be configurable by the bot admin. e.g. if it's less than 2 minutes then wait, otherwise throw externalhosterror

**Comment by viceice** [comment]
sounds good

**Comment by rarkins** [comment]
I wonder if `got` can control precisely this (including the max retry)

**Comment by haldunanil** [comment]
Checking in here to see if this is being worked on/needs to be worked on? I'm also having the same issue in parallelized runs in CI where I run renovate in parallel using one repo/job.

**Comment by rarkins** [comment]
The labels accurately reflect the issue state - the solution is lacking requirements, and is in triage state. You're welcome to assist both with proposing a solution which can be accepted as well as any implementation - thanks

**Comment by felipecrs** [comment]
Maybe it is a different request, but GitHub docs mentions that "you should not retry your request until after the time specified by the `x-ratelimit-reset` header" and also mentions severe penalties for when you are on a secondary rate limit and you continue to make requests when you are rate limited.

https://docs.github.com/pt/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit

My point is: Renovate should stop calling the GitHub API after the first 403/429 response it gets. Whatever call to GitHub API could be catched earlier and just return an error from that function instead if within the rate limit (the call itself would fail if tried anyway). Otherwise, "If your request continues to fail due to a secondary rate limit, wait for an exponentially increasing amount of time between retries, and throw an error after a specific number of retries."

I'm afraid my installation reached the GitHub secondary rate limit already (whose status can't be consulted), and whoever knows when such exponential retry time has gone. I will contact GitHub support to see if I can consult my secondary rate limit status.

Still, I think it'd be a great improvement to Renovate. What do you think?

**Reply by HonkingGoose** [reply]
> My point is: Renovate should stop calling the GitHub API after the first 403/429 response it gets. Whatever call to GitHub API could be catched earlier and just return an error from that function instead if within the rate limit (the call itself would fail if tried anyway). Otherwise, "If your request continues to fail due to a secondary rate limit, wait for an exponentially increasing amount of time between retries, and throw an error after a specific number of retries."

Renovate now has a [`maxRetryAfter`](https://docs.renovatebot.com/configuration-options/#maxretryafter) config option. Here's the full quote from the docs:

> ### maxRetryAfter
>
> Maximum retry-after header value to wait for before retrying a failed request.
>
> Name | Value
> --- | --- 
> type | integer
> default | 60
> parent | hostRules
> 
> A remote host may return a `4xx` response with a `Retry-After` header value, which indicates that Renovate has been rate-limited.
> Renovate may try to contact the host again after waiting a certain time, that's set by the host.
> By default, Renovate tries again after the `Retry-After` header value has passed, up to a maximum of 60 seconds.
> If the `Retry-After` value is more than 60 seconds, Renovate will abort the request instead of waiting.
> 
> You can configure a different maximum value in seconds using `maxRetryAfter`:
> 
> ```json
> {
>   "hostRules": [
>     {
>       "matchHost": "api.github.com",
>       "maxRetryAfter": 25
>     }
>   ]
> }
> ```

If I'm reading the docs right, if you set `maxRetryAfter = 120` or to any value _higher_ than `60` you force Renovate to abort the request. Does that help you? It does feel a bit "hacky" to me. 😄

**Reply by felipecrs** [reply]
@HonkingGoose, thanks a lot. I totally missed this part of the documentation. But after reading it, I have a different understanding...

I believe it means that, with `maxRetryAfter: 60`, the value of the `Retry-After` will only be honored if it's at max 60 seconds. Otherwise, request will be cancelled.

Therefore, if I increase to `maxRetryAfter: 120`, the value of the `Retry-After` would be honored if it's at max 120 seconds. Otherwise, request will be cancelled.

In such case, I believe it would make my situation worse because **less** requests would be cancelled.

--- 

That said, according to this very own documentation, I understand that Renovate already honors `Retry-After` and should therefore not make any requests within `Retry-After` when such host returns it.

@rarkins is this something you can confirm?

**Reply by rarkins** [reply]
It already honors

**Reply by felipecrs** [reply]
That's great. Thank you.
